### DISCUSSION TITLE
Suggestion of updates to the naming convention for Cake Add-ins & Modules (future)

### DISCUSSION THREAD

**Discussion by augustoproiete** [cake-build/cake#2985] [Ideas]
_It might be too late in the game for this and/or might not be worth the effort given the sheer amount of existing add-ins and modules, but thought I would put the suggestion here for discussion just in case_:

The naming convention for add-ins is described on [PR #1352](https://github.com/cake-build/website/pull/1352/files) which officially documents the convention that have been used for years and can be seen on the [many add-ins published on nuget.org](https://www.nuget.org/packages?packagetype=&sortby=created-desc&q=cake.&prerel=false).

The convention states that Cake add-ins should be named:

> _Cake.xxx, where xxx is a meaningful and unique name that describes what the addin does_" which works well.

The naming convention for Cake Modules seems to be similar, just appending a `.Module` suffix based on the [packages on nuget.org](https://www.nuget.org/packages?q=cake+module) today.

Thus, as an example:
- `Cake.Npm` is meant to be used as `#addin` in a Cake script
- `Cake.Npm.Module` is meant to be used as a `#module` in a Cake script

This is nice and has been working well for years... I'm thinking it would be _niceR_ if the convention in the future followed a more namespace-like structure with a more clear separation between add-ins and modules, and at the same time grouping the Cake add-ins together, and the Cake Modules together. e.g.

- `Cake.Addins.xxx`
- `Cake.Modules.xxx`

Where `xxx` is a meaningful and unique name that describes what the add-in or module does.

Updating the example above:
- `Cake.Addins.Npm` is meant to be used as `#addin` in a Cake script
- `Cake.Modules.Npm` is meant to be used as a `#module` in a Cake script

---

The main advantages I think would be:

- Easier to search add-ins and modules, filtering out Cake components, and the kind of add-in desired
- Easier to distinguish Cake add-ins and modules from other components of Cake (e.g. Cake.Common, Cake.CoreCLR, etc.)
- Allow the Cake team to reserve the [ID prefix `Cake.*` to official components only](https://docs.microsoft.com/en-us/nuget/nuget-org/id-prefix-reservation) moving foward, and leave the namespaces `Cake.Addins.*` and `Cake.Modules.*` open for community contributions

**Comment by pascalberger** [comment]
I see the reasoning behind it, but I don't think there's a real chance to make such a switch nowadays, since it would require all addin authors to change their repo & solutions and deprecate NuGet packages. There's no chance to be 100% successful with this, and having half of the addins (which would be a good number) switched to a new system, while the other half is still using old naming conventions would lead to more confusion than it would help. 

As comparison you can take the numbers of addins updated when we break the addin API version. This is a simple change (update Cake reference) for which we even automatically generate PRs for all addins, but we never got higher than 75% percentages of known addins using latest version (see https://github.com/cake-contrib/Home/blob/master/Audit.md#progress).

**Comment by gep13** [comment]
@augustoproiete I have to agree with @pascalberger here.  While it would be great to have this, I think it is fair to say that the horse has already bolted out the stable on this one.  Rather than introduce a new convention, I think we would continue with the convention that we have in place.

**Comment by devlead** [comment]
Also nuget package ids are immutable,  changing names of all would result in having hundreds of obsolete packages just sitting there, which would be confusing for users and a support pain for maintainers.
